src/services/adapters/notification-adapter/notification.payload.builder.ts

```diff
@@ -506,51 +505,87 @@ export class NotificationPayloadBuilder {
     return payload;
   }
 
-  private async getUserContributorPayloadOrFail(
-    userId: string
+  private async getContributorPayloadOrFail(
+    contributorID: string
   ): Promise<ContributorPayload> {
-    const user = await this.entityManager.findOne(User, {
-      where: [
-        {
-          id: userId,
-        },
-        {
-          nameID: userId,
-        },
-      ],
+    const contributor = await this.getContributor(contributorID, {
       relations: {
         profile: true,
       },
     });
 
-    if (!user || !user.profile) {
+    if (!contributor || !contributor.profile) {
       throw new EntityNotFoundException(
-        `Unable to find User with profile for id: ${userId}`,
+        `Unable to find Contributor with profile for id: ${contributorID}`,
         LogContext.COMMUNITY
       );
     }
 
-    const userURL = await this.urlGeneratorService.createUrlForUserNameID(
-      user.nameID
+    const userURL = await this.urlGeneratorService.createUrlForContributor(
+      contributor
     );
     const result: ContributorPayload = {
-      id: user.id,
-      nameID: user.nameID,
+      id: contributor.id,
+      nameID: contributor.nameID,
       profile: {
-        displayName: user.profile.displayName,
+        displayName: contributor.profile.displayName,
         url: userURL,
       },
     };
     return result;
   }
 
+  private async getContributor(
+    contributorID: string,
+    options?: FindOneOptions<IContributor>
+  ): Promise<IContributor | null> {
+    let contributor: IContributor | null;
+    if (contributorID.length === UUID_LENGTH) {
+      contributor = await this.entityManager.findOne(User, {
+        ...options,
+        where: { ...options?.where, id: contributorID },
+      });
+      if (!contributor) {
+        contributor = await this.entityManager.findOne(Organization, {
+          ...options,
+          where: { ...options?.where, id: contributorID },
+        });
+      }
+      if (!contributor) {
+        contributor = await this.entityManager.findOne(VirtualContributor, {
+          ...options,
+          where: { ...options?.where, id: contributorID },
+        });
+      }
+    } else {
+      // look up based on nameID
+      contributor = await this.entityManager.findOne(User, {
+        ...options,
+        where: { ...options?.where, nameID: contributorID },
+      });
+      if (!contributor) {
+        contributor = await this.entityManager.findOne(Organization, {
```

Comment:
Do we have unique ids/nameIDs for all entities? is it possible a name ID of Org to match the nameID of VC but the logic to retrieve the ORG first event if you're looking for the VC?

Reply:
The users, orgs + VCs are all in different namespaces i.e. you could have a user with nameID "john" and also an org and a vc with the same nameID.

So yes this logic has an issue. Can you please raise a separate issue for this one?

Reply:
You are right, this is potentially error-prone. If we have baseContributor, basically nameID should be in its scope, and it is not at the moment.

Reply:
Added a toDo - https://app.zenhub.com/workspaces/alkemio-development-5ecb98b262ebd9f4aec4194c/issues/gh/alkem-io/server/4126 - to fix, and referenced it in the code.